README.md
# [template.js](https://github.com/yanhaijing/template.js)

[![](https://img.shields.io/badge/Powered%20by-jslib%20base-brightgreen.svg)](https://github.com/yanhaijing/jslib-base)
[![license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/yanhaijing/template.js/blob/master/LICENSE)
[![CI](https://github.com/yanhaijing/template.js/actions/workflows/ci.yml/badge.svg?branch=v3)](https://github.com/yanhaijing/template.js/actions/workflows/ci.yml)

English | [简体中文](./README.zh-CN.md)

A javascript template engine that is simple, easy to use, and supports webpack, rspack, vite, esbuild, rollup, parcel, browserify, fis, and gulp.

## Features

- Native JS syntax, template parsing, compilation, and rendering
- Supports all browsers and Node, supports TypeScript
- Precompilation supports mainstream packaging tools
- Custom configurations, decorators, and functions
- Data filtering support
- Exception capture functionality
- Subtemplates
- Sandbox mode
- Supports JIT and AOT compilation modes
- Provides CLI integrated tools
- Provides support for editor plugins

## User Guide

template.js is a better way than concatenating strings, Click to use [codesandbox](https://codesandbox.io/s/template-js-demo-6xgjxw?file=/src/index.mjs) for a quick demo

Template example

```
<ul>
    <%for(var i = 0; i < list.length; i++) {%>
        <li><%:=list[i].name%></li>
    <%}%>
</ul>
```

Data example

```js
const data = {
  list: [{ name: 'yan' }, { name: 'haijing' }],
};
```

Render output

```html
<ul>
  <li>yan</li>
  <li>haijing</li>
</ul>
```

Editor plugins supported by template.js

| Editor   | Plugin                                                                                            |
| -------- | ------------------------------------------------------------------------------------------------- |
| Vscode   | [highlight plugin](https://marketplace.visualstudio.com/items?itemName=yanhaijing1234.templatejs) |
| Sublime  | [highlight plugin](https://packagecontrol.io/packages/templatejs)                                 |
| Atom     | [highlight plugin](https://atom.io/packages/language-templatejs)                                  |
| WebStorm | TODO                                                                                              |

Quickly initialize a project using CLI tools

```bash
$ npx @templatejs/cli new myapp
# Choose the corresponding platform
# ❯ webpack
#   rspack
#   vite
#   rollup
#   esbuild
#   parcel2
#   parcel1
#   fis3
#   browserify
#   gulp
#   browser
```

If you already have a project, you can choose a corresponding plugin, template.js supports different usage methods

| Platform              | Plugin                                                                                                                                            |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| webpack / rspack      | [template-loader](https://github.com/yanhaijing/template.js/blob/master/packages/template-loader)                                                 |
| vite / rollup         | [rollup-plugin-templatejs](https://github.com/yanhaijing/template.js/blob/master/packages/rollup-plugin-templatejs)                               |
| esbuild               | [esbuild-plugin-templatejs](https://github.com/yanhaijing/template.js/blob/master/packages/esbuild-plugin-templatejs)                             |
| parcel2               | [parcel-transformer-template](https://github.com/yanhaijing/template.js/blob/master/packages/parcel-transformer-template)                         |
| parcel1               | [parcel-plugin-template](https://github.com/yanhaijing/template.js/blob/master/packages/parcel-plugin-template)                                   |
| fis                   | [fis-parser-template](https://github.com/yanhaijing/template.js/blob/master/packages/fis-parser-template)                                         |
| browserify            | [browserify-templatejs](https://github.com/yanhaijing/template.js/blob/master/packages/browserify-templatejs)                                     |
| gulp                  | [gulp-templatejs](https://github.com/yanhaijing/template.js/blob/master/packages/gulp-templatejs)                                                 |
| Native web & Node.js  | [template_js](https://github.com/yanhaijing/template.js/blob/master/packages/template)                                                            |
| Unsupported platforms | You can write a plugin yourself, please see [@templatejs/precompiler](https://github.com/yanhaijing/template.js/blob/master/packages/precompiler) |

Other packages summary

- [@templatejs/runtime](https://github.com/yanhaijing/template.js/blob/master/packages/runtime) The template.js runtime, providing runtime support for the rendering function generated by precompiler.
- [@templatejs/parser](https://github.com/yanhaijing/template.js/blob/master/packages/parser) The template.js template parser.
- [template-vscode](https://github.com/yanhaijing/template-vscode) The template.js vscode plugin.

## Developer Guide

This project uses lerna to manage multiple plugins. Common lerna commands are as follows, note that the npx prefix cannot be left out:

```bash
$ npx lerna init # Initialize
$ npx lerna create @templatejs/parser # Create a package
$ npx lerna add yargs --scope=@templatejs/parser # Install package dependencies
$ npx lerna list # List all packages
$ npx lerna bootstrap # Install all dependencies
$ npx lerna link # Create all soft links
$ npx lerna changed # List the packages to be updated next time when using lerna publish
$ npx lerna publish # Will tag, upload git, upload npm
```

Release steps, modify changelog

```bash
$ yarn lint
$ yarn build
$ yarn test
$ npx lerna publish
```

Release without tagging, suitable for test package.

```bash
$ yarn lint
$ yarn build
$ yarn test
$ lerna version --no-git-tag-version
$ npx lerna publish from-package --dist-tag next
```

## Contributors List

[contributors](https://github.com/yanhaijing/template.js/graphs/contributors)

## Changelog

[CHANGELOG.md](https://github.com/yanhaijing/template.js/blob/master/CHANGELOG.md)

## Planned Feature List

[TODO.md](https://github.com/yanhaijing/template.js/blob/master/TODO.md)

## Who is Using

- [Baidu Zhidao](http://zhidao.baidu.com/)
- [Baidu Jingyan](http://jingyan.baidu.com/)
- [Baidu Shifu](http://shifu.baidu.com/)
- Meituan Waimai
- Yuanfudao
- [58 Financial](https://npm.taobao.org/package/jr58)
- Unisplendour
- Uxin

To learn about who is using this, [click here](https://github.com/yanhaijing/template.js/issues/6).

## Related Links

- [BaiduTemplate](http://tangram.baidu.com/BaiduTemplate/)
- [artTemplate](https://github.com/aui/artTemplate/)
- [Juicer](https://github.com/PaulGuo/Juicer)
- [handlebarsjs](http://handlebarsjs.com/)
- [PUG(Originally Jade)](https://pugjs.org/api/getting-started.html)
- [EJS](https://ejs.co/)
- [Only 20 lines of Javascript code! A step-by-step guide to creating a webpage template engine](https://juejin.im/entry/56258da860b294bcf7955883)


CHANGELOG.md
# 变更日志

## 3.1.4 / 2023-11-19

- browserify-templatejs 插件
  - feat: 去掉 expression 参数，现在自动添加 require @templatejs/runtime
- esbuild-plugin-templatejs 插件
  - feat: 去掉 expression 参数，现`在自动添加 import @templatejs/runtime
  - fix: tplName 在windows下显示不正确
- gulp-templatejs 插件
  - feat: 去掉 expression 参数，现在自动添加 require @templatejs/runtime
  - fix: Buffer() is deprecated
- parcel-transformer-template 插件
  - feat: 去掉 expression 参数，现在自动添加 import @templatejs/runtime
  - fix: tplName 只显示文件名，不再显示全路径
- rollup-plugin-templatejs 插件
  - fix: tplName 在windows下显示不正确
- template-loader 插件
  - feat: 去掉 expression 参数，现在自动添加 require @templatejs/runtime
  - fix: tplName 在windows下显示不正确
  - fix: 不同版本 webpack export defalut 处理错误问题
- fix: 修复打包文件的 banner 信息
- fix: cli template 升级为 3.x
- fix: 修复插件 peerDependencies 为 3.x

## 3.0.0-alpha.0 / 2023-11-19

- feat: 支持 nodejs esm exports
- feat: 支持 sourceMap
- refact: 升级 jslib-base 为 2.3.2
- refact: 升级 @jsmini/extend 0.5.0
- refact: 升级 @jsmini/type 0.11.0
- refact: parcel-plugin-template 改为单独发包，从 yarn workspace 和 lerna publish 移出

## 2.10.0 / 2023-11-18

- rollup-plugin-templatejs 插件
  - feat: 去掉 expression 参数，现在自动添加 import @templatejs/runtime

## 2.9.0 / 2023-10-3

- feat: 添加对 parcel v2 的支持
  - 新增 parcel-transformer-template 插件

## 2.8.3 / 2023-10-2

- fix: 修复 parcel v1 demo

## 2.8.2 / 2023-10-1

- 升级 browserify v17 demo
- 升级 esbuild v0.19 demo
- 升级 parcel v1.12 demo
- 升级 rollup v3 demo
- 升级 vite v4.4 demo

## 2.8.1 / 2023-10-1

- 修复 cli new demo error
- 升级 webpack demo 为 webpack5

## 2.8.0 / 2023-9-18

- 添加对 esbuild 的支持

## 2.7.0 / 2023-9-17

- 添加对 rspack 的支持

## 2.6.0 / 2023-9-16

- 添加对 vite 的支持

## 2.5.0 / 2020-8-6

- 增加@template/cli 工具

## 2.4.2 / 2020-7-5

- 修复未传入属性，会访问到原型链属性的 bug
  - in 判断改为 hasOwnProperty 判断
  - 详细讨论见[#39](https://github.com/yanhaijing/template.js/issues/39)

## 2.4.1 / 2020-7-1

- 修复属性为徦值时，比如 0 false ''，获取不到的问题，[#39](https://github.com/yanhaijing/template.js/issues/39)

## 2.4.0 / 2019-12-15

- 添加对`browserify`的支持

## 2.3.0 / 2019-12-15

- 添加对`parcel`的支持

## 2.2.1 / 2019-12-08

- @parser 修复对 crlf 换行符的支持

## 2.2.0 / 2019-11-18

- 添加预编译插件对`sandbox`模式的支持

## 2.1.0 / 2019-11-17

- 完成`gulp-templatejs`

## 2.0.0 / 2019-11-8

- 完成`@templatejs/parser`
- 完成`@templatejs/runtime`
- 完成`@templatejs/precompiler`
- 完成`template_js`
- 完成`fis-parser-template`
- 完成`template-loader`
- 完成`rollup-plugin-templatejs`

## 2.0 之前的变更日志

- [template.js](https://github.com/yanhaijing/template.js/blob/master/packages/template/CHANGELOG.md)
- [fis-parser-template](https://github.com/yanhaijing/template.js/blob/master/packages/fis-parser-template/CHANGELOG.md)
- [template-loader](https://github.com/yanhaijing/template.js/blob/master/packages/template-loader/CHANGELOG.md)
- [rollup-plugin-templatejs](https://github.com/yanhaijing/template.js/blob/master/packages/rollup-plugin-templatejs/CHANGELOG.md)


test/test.js
/**
 * @file 测试文件
 * @author 颜海镜
 * @date 2016年1月20日 21:55:57
 */

var expect = require('expect.js');
var template = require('../template.js');
var t = template;

describe('template.js', function() {
    this.timeout(1000);

    describe('template', function() {
        it('编译', function() {
            expect(typeof t('')).to.equal('function');
        });

        it('简单html', function() {
            var tpl1 = '<div>123</div>';
            expect(t(tpl1, {})).to.equal(tpl1);
        });

        it('多行html', function() {
            var tpl1 = '<div>123</div>\n123';
            expect(t(tpl1, {})).to.equal('<div>123</div>\n123');
        });

        it('输出简单变量', function() {
            var tpl2 = '<div><%=name%></div>';
            expect(t(tpl2, {name: 123})).to.equal('<div>123</div>');
        });

        it('输出html', function() {
            var tpl2 = '<div><%=html%></div>';
            expect(t(tpl2, {html: '<div id="test">'})).to.equal('<div>&lt;div id=&quot;test&quot;&gt;</div>');
            
            var tpl2 = '<div><%:h=html%></div>';
            expect(t(tpl2, {html: '<div id="test">'})).to.equal('<div>&lt;div id=&quot;test&quot;&gt;</div>');
        });

        it('输出变量', function() {
            var tpl3 = '<div><%:=text%></div>';
            expect(t(tpl3, {text: '<div>'})).to.equal('<div><div></div>');
        });

        it('输出url', function() {
            var tpl3 = '<%:u=url%>';
            expect(t(tpl3, {url: 'http://yanhaijing.com?page=颜海镜'})).to.equal('http://yanhaijing.com?page=%E9%A2%9C%E6%B5%B7%E9%95%9C');
        });

        it('自定义变量', function() {
            var tpl = '<%var name = 123;%><%=name%>';
            expect(t(tpl, {})).to.equal('123');
        });

        it('省略结尾分号不会报错', function() {
            var tpl = '<%var a = 1%><%=a%>';
            expect(t(tpl, {})).to.equal('1');
        });

        it('if语句', function() {
            var tpl = '<%if (1) {%>a<%}%>';
            expect(t(tpl, {})).to.equal('a');
        });

        it('for语句', function() {
            var tpl = '<%for(var i = 0; i < 2; i++) {%>a<%}%>';
            expect(t(tpl, {})).to.equal('aa');
        });

        it('while语句', function() {
            var tpl = '<%var a = 3;while(a--) {%>a<%}%>';
            expect(t(tpl, {})).to.equal('aaa');
        });
         
        it('引用空变量返回空字符串', function() {
            var tpl = '<%=a%>';
            expect(t(tpl, {})).to.equal('');
        });
    });

    describe('template.config', function() {
        it('sTag & eTag', function() {
            var tpl = '<#=name%>';
            t.config({sTag: '<#'});
            expect(t(tpl, {name: 123})).to.equal('123');

            var tpl = '<%=name#>';
            t.config({sTag: '<%', eTag: '#>'});
            expect(t(tpl, {name: 123})).to.equal('123');
        });
            
        it('compress', function() {
            var tpl = '<div>  </div>';
            t.config({sTag: '<%', eTag: '%>', compress: true});
            expect(t(tpl, {})).to.equal('<div> </div>');
        });

        it('测试escape', function() {
            var tpl = '<%=html%>';
            t.config({sTag: '<%', eTag: '%>', compress: false, escape: false});
            expect(t(tpl, {html: '<div>'})).to.equal('<div>');
        });
    });

    describe('template modifier&function', function() {
        it('registerModifier', function() {
            t.registerModifier('uc', function (data) {
                return data.toUpperCase();
            });
            var tpl = '<%:uc="yan"%>';
            expect(t(tpl, {})).to.equal('YAN');
        });

        it('registerFunction', function() {
            t.registerFunction('upperCase', function (data) {
                return data.toUpperCase();
            });
            var tpl = '<%=upperCase("yan")%>';
            expect(t(tpl, {})).to.equal('YAN');
        });
    });
});


test/node_test.js
'use strict';

var grunt = require('grunt');
var t = require('../template.js');

/*
  ======== A Handy Little Nodeunit Reference ========
  https://github.com/caolan/nodeunit

  Test methods:
    test.expect(numAssertions)
    test.done()
  Test assertions:
    test.ok(value, [message])
    test.equal(actual, expected, [message])
    test.notEqual(actual, expected, [message])
    test.deepEqual(actual, expected, [message])
    test.notDeepEqual(actual, expected, [message])
    test.strictEqual(actual, expected, [message])
    test.notStrictEqual(actual, expected, [message])
    test.throws(block, [error], [message])
    test.doesNotThrow(block, [error], [message])
    test.ifError(value)
*/

exports.data_js = {
  setUp: function(done) {
    // setup here if necessary
    done();
  },
  default_options: function(test) {
    //test.expect(2);
    var assert = test;
    
    assert.ok(typeof t('') === 'function', '编译');

    var tpl1 = '<div>123</div>';
    assert.ok(t(tpl1, {}) === tpl1, "简单html");

    var tpl2 = '<div><%=name%></div>';
    assert.ok(t(tpl2, {name: 123}) === '<div>123</div>', '输出简答变量');
    var tpl2 = '<div><%=html%></div>';
    assert.ok(t(tpl2, {html: '<div id="test">'}) === '<div>&lt;div id=&quot;test&quot;&gt;</div>', '输出html');
    var tpl2 = '<div><%:h=html%></div>';
    assert.ok(t(tpl2, {html: '<div id="test">'}) === '<div>&lt;div id=&quot;test&quot;&gt;</div>', '输出html');

    var tpl3 = '<div><%:=text%></div>';
    assert.ok(t(tpl3, {text: '<div>'}) === '<div><div></div>', '输出变量');

    var tpl3 = '<%:u=url%>';
    assert.ok(t(tpl3, {url: 'http://yanhaijing.com?page=颜海镜'}) === 'http://yanhaijing.com?page=%E9%A2%9C%E6%B5%B7%E9%95%9C', '输出url');

    var tpl = '<%var name = 123;%><%=name%>';
    assert.ok(t(tpl, {}) === '123', 'var name= 123; 自定义变量');

    var tpl = '<%var a = 1%><%=a%>';
    assert.ok(t(tpl, {}) === '1', '省略结尾分号不会报错');

    var tpl = '<%if (1) {%>a<%}%>';
    assert.ok(t(tpl, {}) === 'a', 'if语句');

    var tpl = '<%for(var i = 0; i < 2; i++) {%>a<%}%>';
    assert.ok(t(tpl, {}) === 'aa', 'for语句');

    var tpl = '<%var a = 3;while(a--) {%>a<%}%>';
    assert.ok(t(tpl, {}) === 'aaa', 'while语句');

    var tpl = '<%=a%>';
    assert.ok(t(tpl, {}) === '', '引用空变量返回空字符串');
        
    //===========================================
    var tpl = '<#=name%>';
    t.config({sTag: '<#'});
    assert.ok(t(tpl, {name: 123}) === '123', 'sTag');

    var tpl = '<%=name#>';
    t.config({sTag: '<%', eTag: '#>'});
    assert.ok(t(tpl, {name: 123}) === '123', 'eTag');

    var tpl = '<div>  </div>';
    t.config({sTag: '<%', eTag: '%>', compress: true});
    assert.ok(t(tpl, {}) === '<div> </div>', 'compress');

    var tpl = '<%=html%>';
    t.config({sTag: '<%', eTag: '%>', compress: false, escape: false});
    assert.ok(t(tpl, {html: '<div>'}) === '<div>', 'escape');

    //==========================================
    t.registerModifier('uc', function (data) {
        return data.toUpperCase();
    });
    var tpl = '<%:uc="yan"%>';
    assert.ok(t(tpl, {}) === 'YAN', t(tpl, {}) === 'YAN');
    

    t.registerFunction('upperCase', function (data) {
        return data.toUpperCase();
    });
    var tpl = '<%=upperCase("yan")%>';
    assert.ok(t(tpl, {}) === 'YAN', t(tpl, {}) === 'YAN');

    test.done();
  }
};


template.js
/*!
 * template.js v0.7.1 (https://github.com/yanhaijing/template.js)
 * API https://github.com/yanhaijing/template.js/blob/master/doc/api.md
 * Copyright 2015 yanhaijing. All Rights Reserved
 * Licensed under MIT (https://github.com/yanhaijing/template.js/blob/master/MIT-LICENSE.txt)
 */
;(function(root, factory) {
    var template = factory(root);
    if (typeof define === 'function' && define.amd) {
        // AMD
        define('template', function() {
            return template;
        });
    } else if (typeof exports === 'object') {
        // Node.js
        module.exports = template;
    } else {
        // Browser globals
        var _template = root.template;

        template.noConflict = function() {
            if (root.template === template) {
                root.template = _template;
            }

            return template;
        };
        root.template = template;
    }
}(this, function(root) {
    'use strict';
    var o = {
        sTag: '<%',//开始标签
        eTag: '%>',//结束标签
        compress: false,//是否压缩html
        escape: true, //默认输出是否进行HTML转义
        error: function (e) {}//错误回调
    };
    var functionMap = {}; //内部函数对象
    //修饰器前缀
    var modifierMap = {
        '': function (param) {return nothing(param)},
        'h': function (param) {return encodeHTML(param)},
        'u': function (param) {return encodeURI(param)}
    };

    var toString = {}.toString;
    var slice = [].slice;
    function type(x) {
        if(x === null){
            return 'null';
        }

        var t= typeof x;

        if(t !== 'object'){
            return t;
        }

        var c = toString.call(x).slice(8, -1).toLowerCase();
        if(c !== 'object'){
            return c;
        }

        if(x.constructor==Object){
            return c;
        }

        return 'unknown';
    }

    function isObject(obj) {
        return type(obj) === 'object';
    }
    function isFunction(fn) {
        return type(fn) === 'function';
    }
    function isString(str) {
        return type(str) === 'string';
    }
    function extend() {
        var target = arguments[0] || {};
        var arrs = slice.call(arguments, 1);
        var len = arrs.length;
     
        for (var i = 0; i < len; i++) {
            var arr = arrs[i];
            for (var name in arr) {
                target[name] = arr[name];
            }
     
        }
        return target;
    }
    function clone() {
        var args = slice.call(arguments);
        return extend.apply(null, [{}].concat(args));
    }
    function nothing(param) {
        return param;
    }
    function encodeHTML(source) {
        return String(source)
            .replace(/&/g,'&amp;')
            .replace(/</g,'&lt;')
            .replace(/>/g,'&gt;')
            .replace(/\\/g,'&#92;')
            .replace(/"/g,'&quot;')
            .replace(/'/g,'&#39;');
    }
    function compress(html) {
        return html.replace(/\s+/g, ' ').replace(/<!--[\w\W]*?-->/g, '');
    }
    function consoleAdapter(cmd, msg) {
        typeof console !== 'undefined' && console[cmd] && console[cmd](msg);
    }
    function handelError(e) {
        var message = 'template.js error\n\n';

        for (var key in e) {
            message += '<' + key + '>\n' + e[key] + '\n\n';
        }
        message += '<message>\n' + e.message + '\n\n';
        consoleAdapter('error', message);

        o.error(e);
        function error() {
            return 'template.js error';
        }
        error.toString = function () {
            return '__code__ = "template.js error"';
        }
        return error;
    }
    function parse(tpl, opt) {
        var code = '';
        var sTag = opt.sTag;
        var eTag = opt.eTag;
        var escape = opt.escape;
        function parsehtml(line) {
            // 单双引号转义，换行符替换为空格
            line = line.replace(/('|")/g, '\\$1').replace(/\n/g, ' ');
            return ';__code__ += ("' + line + '")\n';
        }
        function parsejs(line) {              
            //var reg = /^(:?)(.*?)=(.*)$/;
            var reg = /^(?:=|(:.*?)=)(.*)$/
            var html;
            var arr;
            var modifier;

            // = := :*=
            // :h=123 [':h=123', 'h', '123']
            if (arr = reg.exec(line)) {
                html = arr[2]; // 输出
                if (Boolean(arr[1])) {
                    // :开头
                    modifier = arr[1].slice(1);
                } else {
                    // = 开头
                    modifier = escape ? 'h' : '';
                }

                return ';__code__ += __modifierMap__["' + modifier + '"](typeof (' + html + ') !== "undefined" ? (' + html + ') : "")\n';
            }
            
            //原生js
            return ';' + line + '\n';
        }

        var tokens = tpl.split(sTag);

        for (var i = 0, len = tokens.length; i < len; i++) {
            var token = tokens[i].split(eTag);

            if (token.length === 1) {
                code += parsehtml(token[0]);
            } else {
                code += parsejs(token[0], true);
                if (token[1]) {
                    code += parsehtml(token[1]);
                }
            }
        }

        return code;
    }
    function compiler(tpl, opt) {
        var mainCode = parse(tpl, opt);

        var headerCode = '\n' + 
        '    var html = (function (__data__, __modifierMap__) {\n' + 
        '        var __str__ = "", __code__ = "";\n' + 
        '        for(var key in __data__) {\n' + 
        '            __str__+=("var " + key + "=__data__[\'" + key + "\'];");\n' + 
        '        }\n' + 
        '        eval(__str__);\n\n';

        var footerCode = '\n' + 
        '        ;return __code__;\n' + 
        '    }(__data__, __modifierMap__));\n' + 
        '    return html;\n';

        var code = headerCode + mainCode + footerCode;
        code = code.replace(/[\r]/g, ' '); // ie 7 8 会报错，不知道为什么
        try {
            var Render = new Function('__data__', '__modifierMap__', code); 
            Render.toString = function () {
                return mainCode;
            }
            return Render;
        } catch(e) {
            e.temp = 'function anonymous(__data__, __modifierMap__) {' + code + '}';
            throw e;
        }  
    }
    function compile(tpl, opt) {
        opt = clone(o, opt);

        try {
            var Render = compiler(tpl, opt);
        } catch(e) {
            e.name = 'CompileError';
            e.tpl = tpl;
            e.render = e.temp;
            delete e.temp;
            return handelError(e);
        }

        function render(data) {
            data = clone(functionMap, data);
            try {
                var html = Render(data, modifierMap);
                html = opt.compress ? compress(html) : html;
                return html;
            } catch(e) {
                e.name = 'RenderError';
                e.tpl = tpl;
                e.render = Render.toString();
                return handelError(e)();
            }            
        }

        render.toString = function () {
            return Render.toString();
        };
        return render;
    }
    function template(tpl, data) {
        if (typeof tpl !== 'string') {
            return '';
        }

        var fn = compile(tpl);
        if (!isObject(data)) {
            return fn;
        }

        return fn(data);
    }

    template.config = function (option) {
        if (isObject(option)) {
            o = extend(o, option);
        }
        return clone(o);
    };

    template.registerFunction = function(name, fn) {
        if (!isString(name)) {
            return clone(functionMap);
        }
        if (!isFunction(fn)) {
            return functionMap[name];
        }

        return functionMap[name] = fn;
    }
    template.unregisterFunction = function (name) {
        if (!isString(name)) {
            return false;
        }
        delete functionMap[name];
        return true;
    }

    template.registerModifier = function(name, fn) {
        if (!isString(name)) {
            return clone(modifierMap);
        }
        if (!isFunction(fn)) {
            return modifierMap[name];
        }

        return modifierMap[name] = fn;
    }
    template.unregisterModifier = function (name) {
        if (!isString(name)) {
            return false;
        }
        delete modifierMap[name];
        return true;
    }

    template.__encodeHTML = encodeHTML;
    template.__compress = compress;
    template.__handelError = handelError;
    template.__compile = compile;
    template.version = '0.7.1';
    return template;
}));
